cdap-api/src/main/java/co/cask/cdap/internal/schedule/trigger/ScheduleTrigger.java

```diff
@@ -0,0 +1,609 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *  
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *  
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *  
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.schedule.trigger;
+
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import co.cask.cdap.api.ProgramStatus;
+import co.cask.cdap.api.app.ProgramType;
+
+import java.util.Arrays;
+import java.util.EnumSet;
+import java.util.Objects;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+
+
+/**
+ * Represents a trigger in a schedule to be added in an app.
+ */
+public abstract class ScheduleTrigger implements Trigger {
+
+  /**
+   * Represents all known trigger types in an app.
+   */
+  public enum Type {
+    TIME,
+    STREAM_SIZE,
+    PARTITION,
+    PROGRAM_STATUS,
+    AND(true),
+    OR(true);
+
+    private final boolean isComposite;
+
+    Type() {
+      this.isComposite = false;
+    }
+
+    Type(boolean isComposite) {
+      this.isComposite = isComposite;
+    }
+
+    /**
+     * Whether the type trigger is a composite, i.e. a trigger that can contains multiple triggers internally.
+     */
+    public boolean isComposite() {
+      return isComposite;
+    }
+  }
+
+  private final Type type;
+
+  private ScheduleTrigger(Type type) {
+    this.type = type;
+  }
+
+  public Type getType() {
+    return type;
+  }
+
+  public abstract void validate();
+
+  /**
+   * Represents a time trigger in an app.
+   */
+  public static class TimeTrigger extends ScheduleTrigger {
```

Comment:
I was following the style of the existing `ProtoTrigger` class. I can separate them in individual files.